fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java

```diff
@@ -70,51 +70,40 @@ public LocalDate generateNextRepaymentDate(final LocalDate lastRepaymentDate, fi
-                        }
-                    }
+                    // Get difference between first and second semi dates when lastRepayment date day == first semi date
+                    // day
+                    difference = loanApplicationTerms.getSecondDateForSemi().getDayOfMonth()
+                            - loanApplicationTerms.getFirstDateForSemi().getDayOfMonth();
+                } else if (lastRepaymentDate.getDayOfMonth() > loanApplicationTerms.getFirstDateForSemi().getDayOfMonth()) {
+                    LocalDate nextMonth = lastRepaymentDate.plusMonths(1).withDayOfMonth(1);
+
+                    // days between the begining of the month and the first semi day
+                    difference = loanApplicationTerms.getFirstDateForSemi().getDayOfMonth() - nextMonth.getDayOfMonth() + 1;
+
+                    // Add remaining days til end of the month to the difference
+                    difference += lastRepaymentDate.lengthOfMonth() - lastRepaymentDate.getDayOfMonth();
+                } else if (lastRepaymentDate.getDayOfMonth() == firstRepaymentPeriodDate.getDayOfMonth()) {
+                    difference = loanApplicationTerms.getFirstDateForSemi().getDayOfMonth() - lastRepaymentDate.getDayOfMonth();
                 }
+
                 dueRepaymentPeriodDate = getRepaymentPeriodDate(loanApplicationTerms.getRepaymentPeriodFrequencyType(),
                         Math.abs(difference), lastRepaymentDate);
-                if (dueRepaymentPeriodDate.getDayOfMonth() < loanApplicationTerms.getFirstDateForSemi().getDayOfMonth()) {
-                    if (lastRepaymentDate.getMonthValue() == 2
-                            && dueRepaymentPeriodDate.getDayOfMonth() < loanApplicationTerms.getSecondDateForSemi().getDayOfMonth()) {
-                        if (isLeapYear) {
-                            dueRepaymentPeriodDate = dueRepaymentPeriodDate.minusDays(1);
-                        } else {
-                            dueRepaymentPeriodDate = dueRepaymentPeriodDate.minusDays(2);
-                        }
-                    } else {
-                        dueRepaymentPeriodDate = dueRepaymentPeriodDate.plusDays(1);
-                    }
-                } else if (dueRepaymentPeriodDate.getDayOfMonth() > loanApplicationTerms.getFirstDateForSemi().getDayOfMonth()
-                        && dueRepaymentPeriodDate.getDayOfMonth() < loanApplicationTerms.getSecondDateForSemi().getDayOfMonth()) {
-                    //
-                    int some = dueRepaymentPeriodDate.getDayOfMonth();
-                    int other = loanApplicationTerms.getFirstDateForSemi().getDayOfMonth();
-                    // diff
-                    int then = some - other;
-                    //
-                    if (then == 1) {
-                        dueRepaymentPeriodDate = dueRepaymentPeriodDate.minusDays(1);
-                    } else {
-                        dueRepaymentPeriodDate = dueRepaymentPeriodDate.minusDays(2);
+
```

Comment:
Isn't this the same condition in line no. 79 above? So, can't this block be moved inside there?